File: src/main/java/wiresegal/silimatics/common/item/ItemLens.java
package wiresegal.silimatics.common.item;

import com.google.common.collect.Lists;
import net.minecraft.client.renderer.color.IItemColor;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.init.SoundEvents;
import net.minecraft.inventory.EntityEquipmentSlot;
import net.minecraft.item.ItemStack;
import net.minecraft.util.ActionResult;
import net.minecraft.util.EnumHand;
import net.minecraft.util.SoundCategory;
import net.minecraft.world.World;
import net.minecraftforge.common.MinecraftForge;
import net.minecraftforge.event.entity.living.LivingEvent;
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent;
import net.minecraftforge.fml.relauncher.Side;
import net.minecraftforge.fml.relauncher.SideOnly;
import org.jetbrains.annotations.Nullable;
import wiresegal.silimatics.api.lens.ILens;
import wiresegal.silimatics.common.lens.*;
import wiresegal.silimatics.common.lib.LibMisc;
import wiresegal.zenmodelloader.client.core.TooltipHelper;
import wiresegal.zenmodelloader.common.core.IItemColorProvider;
import wiresegal.zenmodelloader.common.items.base.ItemMod;

import java.util.List;

public class ItemLens extends ItemMod implements ILens, IItemColorProvider {

    public static final String OCULATOR = LibMisc.MODID + ":oculator";

    public static ILens[] lenses = new ILens[] {
            new LensDull(),
            new LensRashid(),
            new LensBright(),
            new LensHeat(),
            new LensWindstormer(),
            new LensVoidstorm(),
            new LensTrails(),
            new LensBlood(),
            new LensPain(),
            new LensSun(),
            new LensView()
    };

    public ItemLens(String name) {
        super(name, EnumSandType.Companion.getSandTypeNamesFor(name));
        MinecraftForge.EVENT_BUS.register(new EventHandler());
    }

    public static ILens getLensFromStack(ItemStack stack) {
        if(stack != null)
            return lenses[stack.getMetadata() % lenses.length];
        return null;
    }

    @SideOnly(Side.CLIENT)
    @Override
    public void addInformation(ItemStack stack, EntityPlayer playerIn, List<String> tooltip, boolean advanced) {
        List<String> tempTip = Lists.newArrayList();
        lenses[stack.getMetadata() % lenses.length].addTooltip(stack, playerIn, tempTip, advanced);
        if (tempTip.size() > 0)
            TooltipHelper.INSTANCE.tooltipIfShift(tooltip, () -> tooltip.addAll(tempTip));
    }

    @Override
    public ActionResult<ItemStack> onItemRightClick(ItemStack itemStackIn, World worldIn, EntityPlayer playerIn, EnumHand hand) {
        //todo if player is wearing the frame
        playerIn.setGlowing(false);
        worldIn.playSound(playerIn, playerIn.getPosition(), SoundEvents.ITEM_ARMOR_EQUIP_IRON, SoundCategory.PLAYERS, 1f, 1f);
        return super.onItemRightClick(itemStackIn, worldIn, playerIn, hand);
    }

    @Override
    public void onUsingTick(World world, EntityPlayer player, ItemStack stack) {
        lenses[stack.getMetadata() % lenses.length].onUsingTick(world, player, stack);
    }


    public static class EventHandler {
        @SubscribeEvent
        public void onPlayerTick(LivingEvent.LivingUpdateEvent event) {
            if (!(event.getEntity() instanceof EntityPlayer)) return;
            EntityPlayer player = (EntityPlayer) event.getEntity();
            ItemStack headStack = player.getItemStackFromSlot(EntityEquipmentSlot.HEAD);
            if (headStack != null && headStack.getItem() instanceof ItemLensFrames) {
                ItemStack lensStack = ItemLensFrames.Companion.getLensStack(headStack);
                if (lensStack != null && lensStack.getItem() instanceof ILens && ((ILens) lensStack.getItem()).shouldMarkAsOculator(lensStack))
                    player.getEntityData().getCompoundTag(EntityPlayer.PERSISTED_NBT_TAG).setBoolean(OCULATOR, true);
            }
        }
    }

    @Nullable
    @SideOnly(Side.CLIENT)
    @Override
    public IItemColor getItemColor() {
        return (stack, tintIndex) -> EnumSandType.values()[stack.getItemDamage() % EnumSandType.values().length].getGlassColor();
    }
}

File: src/main/java/wiresegal/silimatics/common/item/ItemLens.kt
package wiresegal.silimatics.common.item

import com.google.common.collect.Lists
import com.google.common.collect.Multimap
import net.minecraft.client.renderer.color.IItemColor
import net.minecraft.entity.ai.attributes.AttributeModifier
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.init.SoundEvents
import net.minecraft.inventory.EntityEquipmentSlot
import net.minecraft.item.ItemStack
import net.minecraft.util.ActionResult
import net.minecraft.util.EnumHand
import net.minecraft.util.SoundCategory
import net.minecraft.world.World
import net.minecraftforge.common.MinecraftForge
import net.minecraftforge.event.entity.living.LivingEvent
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent
import net.minecraftforge.fml.relauncher.Side
import net.minecraftforge.fml.relauncher.SideOnly
import wiresegal.silimatics.api.lens.ILens
import wiresegal.silimatics.common.lens.*
import wiresegal.silimatics.common.lib.LibMisc
import wiresegal.zenmodelloader.client.core.TooltipHelper
import wiresegal.zenmodelloader.common.core.IItemColorProvider
import wiresegal.zenmodelloader.common.items.base.ItemMod
import wiresegal.silimatics.common.item.ItemLensFrames.Companion.getLensStack

class ItemLens(name: String) : ItemMod(name, *EnumSandType.getSandTypeNamesFor(name)), ILens, IItemColorProvider {

    init {
        MinecraftForge.EVENT_BUS.register(EventHandler)
    }

    @SideOnly(Side.CLIENT)
    override fun addInformation(stack: ItemStack, playerIn: EntityPlayer, tooltip: MutableList<String>, advanced: Boolean) {
        val tempTip = Lists.newArrayList<String>()
        getLensFromStack(stack).addTooltip(stack, playerIn, tempTip, advanced)
        if (tempTip.size > 0)
            TooltipHelper.tooltipIfShift(tooltip) { tooltip.addAll(tempTip) }
    }

    override fun addTooltip(stack: ItemStack, playerIn: EntityPlayer, tooltip: MutableList<String>, advanced: Boolean) {
        val tempTip = Lists.newArrayList<String>()
        getLensFromStack(stack).addTooltip(stack, playerIn, tempTip, advanced)
        if (tempTip.size > 0)
            TooltipHelper.tooltipIfShift(tooltip) { tooltip.addAll(tempTip) }
    }

    override fun onUsingTick(world: World, player: EntityPlayer, stack: ItemStack) {
        getLensFromStack(stack).onUsingTick(world, player, stack)
    }

    override fun shouldMarkAsOculator(stack: ItemStack): Boolean {
        return getLensFromStack(stack).shouldMarkAsOculator(stack)
    }

    override fun addAttributes(slot: EntityEquipmentSlot, stack: ItemStack, modifiers: Multimap<String, AttributeModifier>) {
        getLensFromStack(stack).addAttributes(slot, stack, modifiers)
    }

    @SideOnly(Side.CLIENT)
    override fun getItemColor(): IItemColor? {
        return IItemColor { stack, tintIndex -> EnumSandType.values()[stack.itemDamage % EnumSandType.values().size].glassColor }
    }

    companion object {

        object EventHandler {
            val OCULATOR = LibMisc.MODID + ":oculator"

            @SubscribeEvent
            fun onPlayerTick(event: LivingEvent.LivingUpdateEvent) {
                if (event.entity !is EntityPlayer) return
                val player = event.entity as EntityPlayer
                val headStack = player.getItemStackFromSlot(EntityEquipmentSlot.HEAD)
                if (headStack != null && headStack.item is ItemLensFrames) {
                    val lensStack = headStack.getLensStack()
                    if ((lensStack.item as ILens).shouldMarkAsOculator(lensStack))
                        player.entityData.getCompoundTag(EntityPlayer.PERSISTED_NBT_TAG).setBoolean(OCULATOR, true)
                }
            }
        }

        var lenses = arrayOf(LensDull(), LensRashid(), LensOculator(),
                LensFirebringer(), LensWindstormer(), LensVoidstormer(), LensTracker(),
                LensWarrior(), LensTorturer(), LensShocker(), LensCourier())

        fun getLensFromStack(stack: ItemStack): ILens {
            return lenses[stack.metadata % lenses.size]
        }
    }
}
